feat: EXPOSED-755 Introduce Exposed Gradle Plugin for schema migrations

nomisRev opened the pull request:
Introduces the Exposed Migration Gradle Plugin to simplify SQL migration generation using Exposed table definitions. The plugin supports TestContainers for database testing, Flyway for integration, and a highly configurable extension. Includes unit tests for plugin functionality and task behavior.


#### Description

**Summary of the change**: Introduces the Exposed Migration Gradle Plugin to simplify SQL migration generation using Exposed table definitions. The plugin supports TestContainers for database testing, Flyway for integration, and a highly configurable extension. Includes unit tests for plugin functionality and task behavior.

**Detailed description**:
- **What**: This PR introduces a new module with a Gradle plugin that wraps the migration module.
- **Why**: Automatic migration script generation through Gradle with 0 boilerplate.
- **How**: New functionality, no impact on the existing codebase or features.

---

#### Type of Change

Please mark the relevant options with an "X":
- [X] New feature

Affected databases:
- [X] MariaDB
- [X] Mysql5
- [X] Mysql8
- [X] Oracle
- [X] Postgres
- [X] SqlServer
- [X] H2
- [ ] SQLite

#### Checklist

- [X] Unit tests are in place
- [X] The build is green (including the Detekt check)
- [X] All public methods affected by my PR has up to date API docs
- [ ] Documentation for my change is up to date

---

#### Related Issues
[EXPOSED-755](https://youtrack.jetbrains.com/issue/EXPOSED-755/Create-a-migration-Gradle-plugin)

nomisRev committed:
feat: Add Exposed Migration Gradle Plugin

Introduces the Exposed Migration Gradle Plugin to simplify SQL migration generation using Exposed table definitions. The plugin supports TestContainers for database testing, Flyway for integration, and a highly configurable extension. Includes unit tests for plugin functionality and task behavior.

e5l reviewed (approved):
Hey @nomisRev, thank you for the PR!
Looks nice!

Could you drop the `migration` word from the plugin name? I think we also can have the same versioning and include `bom`, but we can do this later

nomisRev committed:
fix: Rename exposed-migration-gradle-plugin to exposed-gradle-plugin to generalize for adding future functionality.

bog-walk reviewed (commented):
Thanks so much for putting all the work into this @nomisRev ! Left some comments.

Two extra thoughts I had when reviewing:

- I saw that the original module name was changed to drop 'migration'. Is the plan to do the same for the package -> 'org.jetbrains.exposed.plugin'? Or swap the order?
- We might need to adjust the tests. Exposed is setup to run all tests on at least 1 db, so I believe that all your tests are being ignored right now by test tasks. And looking on TC, I can't find any. Depending on the approach, it most likely needs to be setup at minimum like `spring-transaction` (to ensure all new tests run, but only once & only during H2_V2 task):
    - `testImplementation()` needs to be swapped out
    - Then in `TestDbDsl.kt/TestDb`, the new module needs to be filtered out from any task that isn't running H2_V2. Similar to how `val ignoresSpringTests` is handled.

Jdbarad commented:
It's good to have this feature. Is there way to fast track this feature?

bog-walk commented:
Hi @Jdbarad There's no way to fast track it per se. But I can assure you of if its current phase: The branch is currently being migrated over & updated and we're revisiting certain aspects, like particularly script naming logic.

Here's the [YT ticket](https://youtrack.jetbrains.com/issue/EXPOSED-755/Create-a-migration-Gradle-plugin) to track for future updates.

bog-walk committed:
feat: Introduce Exposed migration plugin

- Migrate to Exposed 1.+
- Rename all packages to match Exposed 1.+ convention
- Migrate to TestContainers 2.+
- Bumo JDK to 17 and fix issues
- Adjust tests so that they are only run once (on H2) since not db-specific

bog-walk committed:
feat: Introduce Exposed migration Gradle plugin

- Add README
- Change names to reflect a more general plugin with migration extensions as feature
- Add constants where possible
- Synchronize properties and parameters
- Add KDocs
- Add configuration property for description casing
- Add file version format that uses major + timestamp

bog-walk commented on exposed-gradle-plugin/src/main/kotlin/org/jetbrains/exposed/v1/gradle/plugin/GenerateMigrationsWorker.kt:
```diff
@@ -0,0 +1,230 @@
+    }
+
+    @Suppress("NestedBlockDepth")
+    fun findHighestVersion(migrationsDirectory: File): (Int) -> String {
```
Hello, we are definitely exploring alternative version support.
So far we've already added support for `<prefix><major>_<timestamp><separator><description><extension>`, like `V003_20250409162321__create_users_table.sql`.

But I did wonder if a straight timestamp only format would also be good, with support for `<prefix><timestamp><separator><description><extension>`, like `Version20250409162321__create_users_table.sql`.

It won't be an issue to add, but we'll keep considering the most common formats.

bog-walk commented on exposed-gradle-plugin/build.gradle.kts:
```diff
@@ -0,0 +1,75 @@
+        jvmArgs("-XX:MaxPermSize=256m")
+    }
+
+    // Needed for JDK 17 + ProjectBuilder; otherwise first test always fails with 'Could not inject synthetic classes'
+    // https://github.com/gradle/gradle/issues/18647
+    jvmArgs("--add-opens=java.base/java.lang=ALL-UNNAMED", "--add-opens=java.base/java.util=ALL-UNNAMED")
```
Not sure how much of a hack this is, but it's the only solution I could find to this issue.

bog-walk commented on exposed-gradle-plugin/src/main/kotlin/org/jetbrains/exposed/v1/gradle/plugin/GenerateMigrationsWorker.kt:
```diff
@@ -0,0 +1,230 @@
+    }
+
+    private fun URLClassLoader.getClassesInPackage(packageName: String): Sequence<KClass<*>> =
+        getResources(packageName.replace('.', '/')).asSequence().flatMap { resource ->
```
These are old questions from the first review, I'm just leaving them unresolved because I want to try them myself.

bog-walk commented:
@e5l @obabichevjb Here's a breakdown of what's changed since the last time this was reviewed:

- Migrated to Exposed 1.+ (means all packages were changed to match new convention)
- Migrated to TestContainers 2.+
- Bumped JDK to 17/11, so it's now at the same minimum level as all other modules
- Added a [README](https://github.com/JetBrains/Exposed/pull/2451/changes#diff-786559496e09cb5db48a54e42d6ab399585ad97eb6abfde0dc690d953d8ef00b) <-- _highly suggest reading that first to see how the plugin works_
- Changed a lot of names to make it more a general Exposed Gradle plugin with a migration feature task
- Refactored logic for building file descriptions from SQL statements & added more use cases (but not all, as seen by unresolved comments above)
- Added extensions property so user can configure casing of these file descriptions
- Added support for scripts that use timestamps in their file name (only 1 format though, could probably use another)
- Adjusted tests since they don't currently need a db, so they now only run on H2

Here is what I think next steps should be:

- Decide on end-goal of this plugin. Meaning, should it be like the Ktor Gradle plugin, with future room for multiple features? Or should it remain for migration only?
- Decide if the current defaults are what we want. For example, the fallback is either `V#__description.sql` if no scripts found or whatever the final found file looks like (in the edge case that a user has various script formats)?? Seems too non-deterministic.
- Add more extension properties, like an excludeTablesPackage and maybe (if possible?) a versionFormat property?
- Figure out the cleanest way to add integration tests so I can stop testing only locally 📖

e5l reviewed (commented):
@bog-walk, looks nice, thank you for the update! 
- General Exposed plugin with migrations
- the timestamp format looks good (we can fallback to the v* using config flag)
- it's ok to add more extensions on request with next minor release

bog-walk committed:
feat: Introduce Exposed Gradle plugin with migration feature

- Introduce root ExposedExtension that is extension aware
- Nest existing MigrationsExtension
- Clean up version logic and introduce VersionFormat with more timestamp support
- Add integration tests

bog-walk commented:
@e5l @obabichevjb PR has been updated with the following changes:

- Added new `ExposedExtension` to act as the root/parent plugin extension. The original `MigrationsExtension` is now nested in that.
- Added new configuration property `fileVersionFormat` so user can clearly define how file version is generated
    - Supported by new `VersionFormat` enum
    - Default if not configured is `VersionFormat.TIMESTAMP_ONLY`
    - This also means that existing scripts are only checked if we need to determine the next highest major version.
    - There is no longer any issue with last detected file (from mixed file directory) winning.
 - Added new command line argument `--filename` so user could choose to pass in explicit filename & override our internal name generation logic
    - If they use this, it of course means only a single script file (with that name) would be generated for all statements
- Cleaned up TestContainers to use more appropriate image names & make it easier to update in future
- Added new integration tests

Relevant changes have been updated in README

bog-walk committed:
feat: Introduce Exposed Gradle plugin with migration feature

- Remove printlns
- Fix issue with tables that reference either
- Cover case when statement has keyword between CREATE and INDEX

e5l reviewed (changes requested):
I have tried the plugin on some small sample and opened the pr: https://github.com/nomisRev/Exposed/pull/13


## Blocking 
- PE on non-object tables,
-  MAJOR_MINOR misreads timestamps (reproduced by scenario 7 in the sample), 
- missing compileKotlin dependency (Gradle 9 fails validation)

## Non-blocking 
- task caching
- extractIndexInfo searches wrong string,
- tests write to real resources dir

and several nits. @bog-walk, could you please check?

bog-walk committed:
feat: Introduce Exposed Gradle plugin with migrations extension

- Apply review feedback:
- Fix bug where timestamp version is detected as major version
- Fix issue with compile-time classpath resolution
- Fix issue with user-defined filename created outside directory
- Enable description for add/drop of columns
- Allow use of TestContainer images that don't use tag

bog-walk commented:
@e5l Applied all changes from sample PR feedback except the following:

- Integration tests for TestContainers (still figuring out if this is possible via CI)

Works on sample projects on my end now 🎉 No need for `dependsOn("compileKotlin")` in `build.gradle.kts`.

**Note:** Some of the sample expectations are not 100% possible.

- For example, after introducing the foreign key to `Users`, it assumes single statement scripts. The generated script will contain other statements, like dropping an index. This behavior existed before 1.0, so I need to log an issue and figure out if it's a bug or expected.
- It also expects that migration scripts will include `DROP TABLE` statements. This is not possible when the `Cities` class has been deleted, as it will not appear in the package search and therefore not be passed to `MigrationUtils`. I found it initially strange that we have logic for a statement that Exposed will not generate, but I left it in in the event the user somehow adds manual drop statements to the logic. Not 100% if it's actually possible, but I saw no harm in leaving it in.

bog-walk committed:
feat: Introduce Exposed Gradle plugin with migrations extension

- Make sure logged message on task execution only logs distinct file names
- Update README with info about file descriptors

obabichevjb commented on exposed-gradle-plugin/src/test/kotlin/org/jetbrains/exposed/v1/gradle/plugin/migrations/StatementToFileDescriptionGeneratorTest.kt:
```diff
@@ -0,0 +1,334 @@
+class StatementToFileDescriptionGeneratorTest {
+
+    @Test
+    fun `test CREATE TABLE statements`() {
```
Minor: if I'm right we stopped to use `such names for tests` and started to use `suchNamesForTests` because of CI logging, but I'm not sure already that it's still a case

obabichevjb commented on exposed-gradle-plugin/src/main/kotlin/org/jetbrains/exposed/v1/gradle/plugin/GenerateMigrationsWorker.kt:
```diff
@@ -0,0 +1,233 @@
+
+        // Get the canonical path (resolves ../, and other edge cases)
+        val canonicalPath = requestedFile.canonicalPath
+        if (!canonicalPath.startsWith(baseDir.path)) {
```
Minor: is it intended to use `baseDir.path` rather than `baseDir.canonicalPath`?

bog-walk replied:
Hmm I assumed it would be ok given that `baseDir.canonicalPath == File(parentPath.canonicalPath)`.
If you expect it might miss cases, please let's open a ticket and address them.

obabichevjb commented on exposed-gradle-plugin/src/main/kotlin/org/jetbrains/exposed/v1/gradle/plugin/DescriptionGenerator.kt:
```diff
@@ -0,0 +1,191 @@
+        }
+
+        // CREATE INDEX or CREATE ??? INDEX statements
+        Regex("^CREATE\\s+(\\w+\\s+)*INDEX\\s+").containsMatchIn(normalizedStatement.uppercase()) -> {
```
Minor: it's not critical for that tool, but usually it's better to create a regex once as a constant and reuse it.

obabichevjb commented on exposed-gradle-plugin/src/main/kotlin/org/jetbrains/exposed/v1/gradle/plugin/GenerateMigrationsParameters.kt:
```diff
@@ -0,0 +1,25 @@
+interface GenerateMigrationsParameters : WorkParameters {
+    var tablesPackage: String
+    var classpathUrls: List<URL>
+    val fileDirectory: DirectoryProperty
```
Probably it would be better to add prefix `migration` for the parameters related to migration file (like `migrationFileDirectory`, `migrationFilePrefix`), otherwise it's quite generic and on the first reading I was trying to predict which file is supposed to be here.

bog-walk replied:
Given their context use, wouldn't that look messy? It's not possible to use them elsewhere in plugin.
```kts
exposed {
    migrations {
        migrationFileDirectory.set()
        migrationFilePrefix.set()
    }
}
```

If you'd still like to consider changing these names, please let's discuss on Monday, so we can change next week.

e5l reviewed (approved):
lgtm

bog-walk committed:
feat: Introduce Exposed Gradle plugin with migrations extension

- Address review feedback

bog-walk committed:
feat: Introduce Exposed Gradle plugin with migrations extension

- Add missing interface KDocs

Jdbarad commented:
Good to hear 😊, finally geadle plugin is ready to use.
